metals/src/main/scala/scala/meta/internal/metals/ConnectionProvider.scala

```diff
@@ -0,0 +1,729 @@
+    resetService: () => Unit,
+    doctor: Doctor,
+    initTreeView: () => Unit,
+    diagnostics: Diagnostics,
+    charset: Charset,
+    buildClient: MetalsBuildClient,
+    bspGlobalDirectories: List[AbsolutePath],
+    bspStatus: bsp.ConnectionBspStatus,
+    mainBuildTargetsData: TargetData,
+)(implicit ec: ExecutionContextExecutorService, rc: ReportContext)
+    extends Indexer(
+      languageClient,
+      executionContext,
+      tables,
+      statusBar,
+      workDoneProgress,
+      timerProvider,
+      indexingPromise,
+      buildData,
+      clientConfig,
+      definitionIndex,
+      referencesProvider,
+      workspaceSymbols,
+      buildTargets,
+      semanticDBIndexer,
+      fileWatcher,
+      focusedDocument,
+      focusedDocumentBuildTarget,
+      buildTargetClasses,
+      userConfig,
+      sh,
+      symbolDocs,
+      scalaVersionSelector,
+      sourceMapper,
+      folder,
+      implementationProvider,
+      resetService,
+    )
+    with Cancelable {
+  import ConnectProvider.connect
+
+  def resolveBsp(): bsp.BspResolvedResult =
+    bspConnector.resolve(buildToolProvider.buildTool)
+
+  protected val bspServers: bsp.BspServers = new bsp.BspServers(
+    folder,
+    charset,
+    languageClient,
+    buildClient,
+    tables,
+    bspGlobalDirectories,
+    clientConfig.initialConfig,
+    userConfig,
+    workDoneProgress,
+  )
+
+  val bspConnector: BspConnector = new BspConnector(
+    bloopServers,
+    bspServers,
+    buildTools,
+    languageClient,
+    tables,
+    userConfig,
+    statusBar,
+    workDoneProgress,
+    bspConfigGenerator,
+    () => bspSession.map(_.mainConnection),
+    () => connect(new CreateSession(true)).ignoreValue,
+    bspStatus,
+  )
+
+  private val bloopInstall: BloopInstall = new BloopInstall(
+    folder,
+    languageClient,
+    buildTools,
+    tables,
+    shellRunner,
+    userConfig,
+  )
+
+  val cancelables = new MutableCancelable
+  var buildServerPromise: Promise[Unit] = Promise[Unit]()
+  val isConnecting = new AtomicBoolean(false)
+
+  override def index(check: () => Unit): Future[Unit] = connect(
+    Index(check)
+  ).ignoreValue
+  override def cancel(): Unit = {
+    cancelables.cancel()
+  }
+
+  object ConnectProvider {
+    def connect[T](config: ConnectConfig[T]): Future[BuildChange] = {
+      val provider =
+        config match {
+          case _: Disconnect => DisconnectProvider
+          case _: Index => IndexProvider
+          case _: ImportBuildAndIndex => ImportBuildAndIndexProvider
+          case _: Connect => ConnectProvider
+          case _: CreateSession => CreateSessionProvider
+          case _: GenerateBspConfigAndConnect =>
+            GenerateBspConfigAndConnectProvider
+          case _: BloopInstallAndConnect => BloopInstallAndConnectProvider
+        }
+      provider.asInstanceOf[ConnectProvider[T, config.type]].run(config.config)
+    }
+
+    private sealed trait ConnectProvider[T, X <: ConnectConfig[T]] {
```

Comment:
Do we need the trait actually? This would only be used above and we can just instead invoke run on each provider

Reply:
No, we don't. You're right that the common interface here might not be too beneficial anyway. Maybe I'll delete the providers and just use methods instead.